Create SwiftABI.md (#139)
WebAssembly/tool-conventions: SwiftABI.md

This document describes the Swift ABI for WebAssembly. As mentioned in README, it's not the only possible Swift ABI. It is the ABI that the clang/LLVM WebAssembly backend is currently using, and any other Swift compiler wishing to be ABI-compatible with it.

Most parts of the ABI follow [the Swift ABI document in the official compiler repository](https://github.com/apple/swift/tree/master/docs/ABI)

This document only describes what is necessary to be treated specially on WebAssembly.

## Swift Calling Convention

Swift calling convention (`swiftcc` attribute in LLVM) is based on C calling convention as described [here](https://github.com/apple/swift/blob/master/docs/ABI/CallingConvention.rst)

Swiftcc on WebAssembly is a little different from swiftcc on other architectures.

> On the non-WebAssembly arch, swiftcc accepts extra tail parameters that are attributed with swifterror or swiftself by a caller at the LLVM IR level. Even if callee doesn't have these parameters at the LLVM IR level, the invocation succeeds by ignoring extra parameters.

Reviewer: ```suggestion
On non-WebAssembly architectures, swiftcc accepts extra tail parameters that are attributed with swifterror or swiftself by a caller at the LLVM IR level. Even if the callee doesn't have these parameters at the LLVM IR level, the invocation succeeds by ignoring extra parameters.
```

But WebAssembly strictly checks that [callee and caller signatures are the same](https://github.com/WebAssembly/design/blob/master/Semantics.md#calls).

> So at WebAssembly level, all swiftcc functions end up extra arguments and all function definitions and invocations explicitly have additional tail parameters to fill swifterror and swiftself.

Reviewer: ```suggestion
So at WebAssembly level, all swiftcc functions end up with extra arguments and all function definitions and invocations explicitly have additional tail parameters to fill swifterror and swiftself.
```

For example, Swift global function `func foo(_ value: Int)` is lowered as `func (param i32 i32 i32)` at WebAssembly level.

> If you export swiftcc function and call it, you maybe need to pass additional parameters to fill them. It's OK to fill the placeholder with any pointer size value.

Reviewer: If a function is exported to JS, then whether extra parameters are strictly needed depends where in the parameter list they are. if they are on the end, then missing params will be added as JS undefined and coerced to i32 as 0.

Reviewer: Also, if the exported function actually does use swiftself/swifterror (i.e. they were not just added on during lowering), then you can't just put any placeholder, correct? Is it possible to export and use these functions?

Author: > if they are on the end, then missing params will be added as JS undefined and coerced to i32 as 0.

I didn't know that spec, is there any documentation? I'll add the link in this doc. 

> Also, if the exported function actually does use swiftself/swifterror (i.e. they were not just added on during lowering), then you can't just put any placeholder, correct? Is it possible to export and use these functions?

yes. In that case, the caller should add these parameters correctly. In theory, it's possible to export and use swiftcc functions. But in general, the functions to be exported to JS are exported with C calling convention.

Reviewer: The wasm JS spec is at https://webassembly.github.io/spec/js-api/index.html
In this doc you could maybe just say that exported swiftcc functions also need the tail parameters, but that if the embedding environment is JS, then the JS conversion rules will supply them.

> If you export swiftcc function and call it, you need the tail parameters. It's OK to fill the placeholder with any pointer size value. But if the embedding environment is JS, then missing params will be added as JS undefined and coerced to i32 as 0, so you don't need to fill them. This JS convention rule is documented at https://webassembly.github.io/spec/js-api/index.html

Reviewer: ```suggestion
If you export swiftcc function and call it, you need the tail parameters. If the function did not have swiftself or swifterror at the IR leve (and placeholders were added during lowering), then a placeholder may be passed any pointer size value. But if the embedding environment is JS, then missing params will be added as JS undefined and coerced to i32 as 0, so you don't need to fill them. This JS convention rule is documented at https://webassembly.github.io/spec/js-api/index.html
```

Reviewer: oops, the suggestion has a typo: "leve" -> "level"

Author: 👍
